### ISSUE TITLE
logging?

### ISSUE THREAD

**Issue by thesps** [fastmachinelearning/hls4ml#589]
It could be useful to have a logging utility for a more verbose interaction with hls4ml. At the moment we print some information about the model during conversion, and some other warnings (invalid RF comes to mind), but we could print a whole lot more and make the whole thing configurable with different detail levels. I think it could help both users and developers, and also with developers debugging users' issues.

One specific place I think it could add a lot of information would be during the model optimization. If each pass provides some information about what condition caused a match, and what exactly it's modifying (e.g. if it's replacing a node) that could help a lot debugging when things go wrong.

e.g. built in `logging`: https://docs.python.org/3/library/logging.html

**Comment by bobbomania** [comment]
Hi, it's been a while, but anyways I think this has been done for the some part in the optimisation module, but there are still some printing throughout the codebase- are there still relevant files to convert to use logging and is this still a worthwhile issue to take up?

**Comment by thesps** [comment]
I would let @JanFSchulte and @bo3z comment, but I would say _yes_ it's still worthwhile to take this up. Essentially anywhere there is a `print` in the Python could be replaced by `logging` with different severity levels, and more messages could be added throughout.

**Comment by bo3z** [comment]
I agree - this was discussed in one of the recent dev meetings. While we haven't converged towards a solution yet, the general conclusion is that logging approaches (logging, print or something third) should be unified. The best approach for now seems to shift to the logging module as the @thesps suggested.

**Comment by bobbomania** [comment]
Ok, I will go through the code-base and familiarise myself to see what can be logged.

Also, as part of my research project, I integrated hls4ml with post-place and route power estimation using Vivado for both io_stream and io_parallel - would that be a useful contribution for hls4ml, could I create an issue for it?

**Comment by vloncar** [comment]
> Also, as part of my research project, I integrated hls4ml with post-place and route power estimation using Vivado for both io_stream and io_parallel - would that be a useful contribution for hls4ml, could I create an issue for it?

that would be a more useful contribution than logging 😉
